src/vision/status_quo/barbara_battles_buffered_streams.md

```diff
@@ -0,0 +1,216 @@
+
+It turns out that, to Rust, a task is kind of a black box with a "poll" function. When the executor thinks a task can make progress, it calls poll. The task itself then delegates this call to poll down to all the other futures that are composed together. In the case of her buffered stream of connections, the stream gets woken up and it would then delegate down the various buffered items in its list.
+
+When it executes `Stream::for_each`, the task is doing something like this:
+
+```rust
+while let Some(work_item) = stream.next().await {
+    process_work_item(database, work_item).await;
+}
+```
+
+The task can only "wait" on one "await" at a time. It will execute that await until it completes and only then move on to the rest of the function. When the task is blocked on the first `await`, it will process all the futures that are part of the stream, and hence the various buffered connections all make progress. 
+
+But once a work item is produced, the task will block on the *second* `await` -- the one that resulted from `process_work_item`. This means that, until `process_work_item` completes, control will *never return to the first `await`*. As a result, none of the futures in the stream will make progress, even if they could do so!
+
+### The fix
+
+Once Barbara understands the problem, she considers the fix. The most obvious fix is to spawn out tasks that will process each work item, like so:
+
+```rust
+async fn do_work(database: &Database) {
+    let work = do_select(database, FIND_WORK_QUERY)?;
+    stream::iter(work.into_iter())
+        .map(|item| do_select(database, work_from_item(item)))
+        .buffered(5)
+        .for_each(|work_item| task::spawn(async move {
+            process_work_item(database, work_item).await
+        })
+        .await;
+}
+```
+
+Spawning a task completes immediately, so the task will quickly move back to blocking on the stream and allowing other connections to make progress. Unfortunately, this change results in a compilation error:
+
+```
+error[E0759]: `database` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
+  --> src/main.rs:8:18
+   |
+8  | async fn do_work(database: &Database) {
+   |                  ^^^^^^^^  --------- this data with an anonymous lifetime `'_`...
+   |                  |
+   |                  ...is captured here...
+   |        .for_each(|work_item| task::spawn(async move {
+   |                              ----------- ...and is required to live as long as `'static` here
+```
+
+"Ah, right," she says, "spawned tasks can't use borrowed data. I wish I had [rayon] or the scoped threads from [crossbeam]." (What Barbara doesn't realize is that spawning wouldn't actually have fixed her problem anyway: the `for_each` combinator would have awaited the resulting `JoinHandle` and hence it would have blocked... but she could have tweaked her program to fix that if she had gotten that far.)
```

Comment:
It's true that the `spawn` used here would not have fixed her problem, but spawning _would_ have fixed her problem if she had put the spawn around `do_select` rather than around `process_work_item`. (Ignoring any issues regarding `'static`)

Reply:
Yes, I should note that in the FAQ,

Reply:
The fix we use most commonly is described below - we spawn inside `do_select` to ensure that the work done there can complete.